### ISSUE TITLE
setenv() not working for PATH

### ISSUE THREAD

**Issue by GlenHertz** [JuliaLang/julia#13265]
Using 0.3.9 and 0.4.0-rc2 it doesn't seem like `setenv` works:

``` Julia
cmd = `ls`
vars = ["/nowhere"] #["PATH=/usr/bin:/bin"]
cmd = setenv(cmd, vars)
@show readall(cmd)  # should fail but doesn't
```

I expect the above to fail on Linux but it seems to use the existing env vars.  I'm confused about what `setenv` does.  If it is a `set` function shouldn't it end with a `!`.  I first thought it would alter the cmd (and didn't assign it to `cmd` but above I fixed that).  Perhaps there should be a `setenv` and `setenv!`.

It would be great if this would work as it is a pain to do this in a Unix shell.

Glen

**Comment by GlenHertz** [comment]
It doesn't make a difference but in the above testcase I meant to use `vars = ["PATH=/nowhere"]`.

**Comment by vtjnash** [comment]
https://github.com/joyent/libuv/issues/122

**Comment by GlenHertz** [comment]
Interesting, I'm not sure if that is the same issues that I'm seeing.  My issue seems to be that `PATH` is not being used at all:

```Julia
dir = "/tmp"
file = "test1234"
abspath = joinpath(dir, file)
open(abspath, "w") do f
	println(f, """
	#!/bin/bash
	echo "Hi"
	""")
end
chmod(abspath, 0o777)
vars = ["PATH=$dir:/bin:/usr/bin"]
cmd = `$file`
cmd = setenv(cmd, vars)
print(readall(cmd))  # should pass
```

In joyent/libuv#122 it seems to say that the PATH will be used on Linux (if provided).  I'm 99% sure this was working in the early 0.3 days.

**Comment by staticfloat** [comment]
This has been fixed; on 1.4.2, the original test case now fails, here's a simple reproducer:

```julia
julia> run(setenv(`true`))
Process(`true`, ProcessExited(0))

julia> run(setenv(`true`, "PATH" => "/blah"))
ERROR: IOError: could not spawn setenv(`true`,["PATH=/blah"]): no such file or directory (ENOENT)
Stacktrace:
 [1] _spawn_primitive(::String, ::Cmd, ::Array{Any,1}) at ./process.jl:99
 [2] #550 at ./process.jl:112 [inlined]
 [3] setup_stdios(::Base.var"#550#551"{Cmd}, ::Array{Any,1}) at ./process.jl:196
 [4] _spawn at ./process.jl:111 [inlined]
 [5] run(::Cmd; wait::Bool) at ./process.jl:439
 [6] run(::Cmd) at ./process.jl:438
 [7] top-level scope at REPL[11]:1
```

Adapting the slightly more complex example above, here's an example proving that `PATH` is being set properly for subprocesses as well (since in the second example, by embedding `#!/bin/bash` as the interpreter and only invoking `echo`, the script may not need `PATH` at all); we create a script that declares `/usr/bin/env` as the interpreter, and tells it to search for `bash`, then we show that if we allow `/bin` onto the `PATH`, everything works:

```julia
julia> mktempdir() do dir
       f = joinpath(dir, "test1234")
       open(f, "w") do io
           write(io, """
           #!/usr/bin/env bash
           echo "Hello, Julia!"
           """)
       end
       chmod(f, 0o777)
       run(setenv(`$f`, "PATH" => "/bin"))
       end
Hello, Julia!
Process(setenv(`/var/folders/xz/tg6x2wt15y70nmr5cb0bq2k00000gn/T/jl_nQwYLk/test1234`,["PATH=/bin"]), ProcessExited(0))
```

Setting `PATH` to `/blah` results in:
```julia
env: bash: No such file or directory
ERROR: failed process: Process(setenv(`/var/folders/xz/tg6x2wt15y70nmr5cb0bq2k00000gn/T/jl_bjvyEC/test1234`,["PATH=/blah"]), ProcessExited(127)) [127]

Stacktrace:
 [1] pipeline_error at ./process.jl:525 [inlined]
 [2] run(::Cmd; wait::Bool) at ./process.jl:440
 [3] run at ./process.jl:438 [inlined]
 [4] (::var"#19#21")(::String) at ./REPL[17]:10
 [5] mktempdir(::var"#19#21", ::String; prefix::String) at ./file.jl:682
 [6] mktempdir(::Function, ::String) at ./file.jl:680 (repeats 2 times)
 [7] top-level scope at REPL[17]:1

```
